Call the filterchain after the callback

andre-doherty opened the pull request:
When the callback logic has been called, call the remaining filter chain 
[https://github.com/bujiio/buji-pac4j/issues/67](https://github.com/bujiio/buji-pac4j/issues/67)

leleuj commented:
That was a bit too easy ;-) You need to update the version in the pom.xml to 3.1.0-SNAPSHOT (as it's a change and bot a bugfix) and update the README as well (here and there). Thanks.

andre-doherty commented:
Ok i see, i didn't know how you were dealing with new milestones.
Ok will do!

Le 4 déc. 2017 18:05, "LELEU Jérôme" <notifications@github.com> a écrit :

> That was a bit too easy ;-) You need to update the version in the pom.xml
> to 3.1.0-SNAPSHOT (as it's a change and bot a bugfix) and update the README
> as well (here and there). Thanks.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bujiio/buji-pac4j/pull/68#issuecomment-349029534>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMI_yn0knVwanPqBziSoM-s0flQPlLYiks5s9CZcgaJpZM4Q075Y>
> .
>

andre-doherty commented:
Should be better this time.
Please let me know if that's ok

leleuj commented:
Great! Thanks

leleuj commented:
@andre-doherty I just tested your change with the demo: https://github.com/pac4j/buji-pac4j-demo and I get a 404 error on the /callback URL: does it work for you?

andre-doherty commented:
Yes, same :(

Looks like there is a slight side-effect, possibly on the way the redirect
is handled
ᐧ

On Tue, Dec 5, 2017 at 5:08 AM, LELEU Jérôme <notifications@github.com>
wrote:

> @andre-doherty <https://github.com/andre-doherty> I just tested your
> change with the demo: https://github.com/pac4j/buji-pac4j-demo and I get
> a 404 error on the /callback URL: does it work for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bujiio/buji-pac4j/pull/68#issuecomment-349189465>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMI_ylFLJkm7RSWeMg-as_Mg6SRC0Mr4ks5s9MHCgaJpZM4Q075Y>
> .
>



-- 
Andre Doherty
Tel : +33 (0) 6 14 18 96 46

leleuj commented:
Can you investigate on this? Thanks

andre-doherty commented:
Well, as far as i understand, when the filter chain has been fully executed the request processor intends to resolve the initial request, not taking into account a 302 status / location header previously set.
The "standard" way to doing redirect would of course to invoke response.sendRedirect(), but this commit the response, so no further way to add some elements in it ...
If the filterchain is broken, the magic does the trick and this time 302/location is used ...

So i see one possible option : 
- add a supplementary filter that would break the filter chain at its end but only when in a callback context, break or even better do a response.sendRedirect("backed url") (it actually works)
- i have to think more about whether this filter should be handled at web.xml level or shiro, the first being straighforward but adding more complexity to the user

But you might have a better idea ?

andre-doherty commented:
Hi,

Another option would be to not to do any magic trick : instead declare a real callback servlet. The servlet would redirect to where the url the callback filter would use in the first place.

This behaviour could be activated as an optional one in the callback filter settings.
It's more natural to me (tricks always backfire :) )

andre-doherty commented:
Well HttpActionAdapter is indeed a fine place to insert some user specific code. 
However as it was hard coded i didnt go into that track.

So yes, if i am able to specify a customHttpAdapter, i will be able to add some extra infos the response (such as a jwt token for instance).
Opening the possibility to override the J2ENopHttpActionAdapter instance would fit the need (just tested it and its ok).
And in this case we could stick to the broker filterchain as the goal would be achieved.

But i don't see quite what you mean by : 

> we could have a special CallbackHttpActionAdapter handling the 302 and 200 HTTP responses (200 for CAS logout).
Can you be more specific ?

leleuj commented:
In fact, I was wrong: the `HttpActionAdapter` is called before the `filterChain` so it cannot do anything usefull unless we imagine retaining the `HttpAction` and "replaying" it after the `filterChain`...

But your proposal makes sense: we could have an `HttpActionAdapter` at the `CallbackFilter` level and if not defined, use the `J2ENopHttpActionAdapter`.

Can you submit a new PR? Thanks

andre-doherty commented:
It's done.

lcconly commented:
I have the same problem with the demo , showing as the /callback Url is 404 not found, when Cas client gets tickets from Server. Is there any way to solve this problem?

leleuj commented:
The demo still uses buji-pac4j v3.0.0 so you cannot encounter this issue. Don't use version 3.1.0-SNAPSHOT for now. We are working on fixing that...

lcconly commented:
So, you mean it can not happen in v3.0.0. Is there any extra code to operate on callbackurl for CAS? Since I follow the introduction to set clients.callbackUrl and set /callbachUrl=callbackFilter, it is still error. So, I wonder it is I miss some steps.

leleuj commented:
@lcconly it should be straightforward. Nothing special required. It works in the demo with the appropriate CAS server URL.

@andre-doherty I replaced your PR with this one: https://github.com/bujiio/buji-pac4j/pull/70 and updated the demo: https://github.com/pac4j/buji-pac4j-demo/commit/6c86c96de27eafcac42a42ada2384c2128cbc17a

andre-doherty commented:
@leleuj : Thank you ! Something went wrong with my git ... Should be ok
next time
ᐧ

On Fri, Dec 8, 2017 at 12:32 PM, LELEU Jérôme <notifications@github.com>
wrote:

> @lcconly <https://github.com/lcconly> it should be straightforward.
> Nothing special required. It works in the demo with the appropriate CAS
> server URL.
>
> @andre-doherty <https://github.com/andre-doherty> I replaced your PR with
> this one: #70 <https://github.com/bujiio/buji-pac4j/pull/70> and updated
> the demo: pac4j/buji-pac4j-demo@6c86c96
> <https://github.com/pac4j/buji-pac4j-demo/commit/6c86c96de27eafcac42a42ada2384c2128cbc17a>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bujiio/buji-pac4j/pull/68#issuecomment-350242348>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMI_yo68rLyuoAjNMbj5bLMqT8N8IMRXks5s-R5QgaJpZM4Q075Y>
> .
>



-- 
Andre Doherty
Tel : +33 (0) 6 14 18 96 46
